### PR TITLE
OCPCLOUD-2010: Add CloudControllerManagerStatus to External platform status

### PR DESCRIPTION
**Pull request by adriengentil**
Add the CloudControllerManagerStatus to the ExternalPlatformStatus struct.
The new field is added behind `TechPreviewNoUpgrade` feature set.

This setting will be set at installation time.

* Enhancement doc update PR (https://github.com/openshift/enhancements/pull/1378)

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Revert so not much to review here, but, do we have an implementation PR that can be reviewed alongside this API change?

**Comment by adriengentil** [comment]
> Revert so not much to review here, but, do we have an implementation PR that can be reviewed alongside this API change?

I'm currently implementing the other part in library-go, I'll link it here once I have a draft.

**Review by elmiko — COMMENTED** [review]
/lgtm

@JoelSpeed ptal

File: config/v1/types_infrastructure.go [hunk 1]
```diff
@@ -230,6 +230,36 @@ const (
...
+	// be installed within the cluster.
+	// https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+	//
+	// When set to "External", new nodes will be tainted as uninitialized when created,
+	// preventing them from running workloads until they are initialized by the cloud controller manager.
+	// When omitted or set to "None", new nodes will be not tainted
+	// and no extra initialization from the cloud controller manager is expected.
+	// +kubebuilder:validation:Enum="";External;None
+	// +optional
+	State CloudControllerManagerState `json:"state"`
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:260]: Do we allow this to be changed post install? Are there any transitions we would like to prevent, eg, should the `External` not be allowed to transition back to `None`?

Comment (elmiko) [reply config/v1/types_infrastructure.go:260]: we have never discussed being able to transit back from CCMs to non-CCMs. given that this field is intended for deployments where there most likely won't be an in-tree provider, i don't think we will allow it to be changed after install.

i'm trying to think about what that would mean, and essentially it would be switching a cluster from using CCMs to using no cloud controllers. which sounds like it might work in theory, but i'm not sure about the transitive effects and if that is supportable for us.

Comment (elmiko) [reply config/v1/types_infrastructure.go:260]: Joel and i discussed this in our standup today, we have some questions about whether this field should be in the spec or status of the infra object. i'd like to get a few more opinions about how we intend this feature to be used, and if it is appropriate for users to change it after installation.

**Comment by elmiko** [comment]
i do see some the values updated in the installer, i'm guessing we will need to do this when adding support for external platform  there, but i'm not sure if this is definitive.

https://github.com/openshift/installer/blob/master/pkg/asset/manifests/infrastructure.go

**Comment by adriengentil** [comment]
/retitle Add CloudControllerManagerStatus to External platform status

**Comment by adriengentil** [comment]
/retitle OCPCLOUD-2010: Add CloudControllerManagerStatus to External platform status

**Comment by elmiko** [comment]
Joel, Adrien, and i had a chat today and we have agreed to add the new fields only to the status section as we won't enable an active migration to CCMs, or from, at this time. the installer will update the status block from the install config values.

File: config/v1/types_infrastructure.go [hunk 2]
```diff
@@ -315,8 +315,44 @@ type PlatformSpec struct {
...
+	// state determines whether or not an external Cloud Controller Manager is expected to
+	// be installed within the cluster.
+	// https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+	//
+	// When set to "External", new nodes will be tainted as uninitialized when created,
+	// preventing them from running workloads until they are initialized by the cloud controller manager.
+	// When omitted or set to "None", new nodes will be not tainted
+	// and no extra initialization from the cloud controller manager is expected.
+	// +kubebuilder:default:=""
+	// +default=""
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:344]: Don't think we need these

File: config/v1/types_infrastructure.go [hunk 3]
```diff
@@ -315,8 +315,44 @@ type PlatformSpec struct {
...
+	// +kubebuilder:validation:Enum="";External;None
+	// +optional
+	State CloudControllerManagerState `json:"state"`
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:347]: Please add a CEL validation to ensure this cannot be mutated once set

File: config/v1/types_infrastructure.go [hunk 4]
```diff
@@ -315,8 +315,44 @@ type PlatformSpec struct {
...
+	CloudControllerManagerExternal CloudControllerManagerState = "External"
+
+	// Cloud Controller Manager is disabled and not expected to be installed.
+	// This value indicates that new nodes should not be tainted
+	// and no extra node initialization is expected from the cloud controller manager.
+	CloudControllerManagerNone CloudControllerManagerState = "None"
+)
+
+// CloudControllerManagerStatus holds the state of Cloud Controller Manager (a.k.a. CCM or CPI) related settings
+type CloudControllerManagerStatus struct {
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:334]: Please add a CEL validation to make sure the state field cannot be removed once it is added

**Review by JoelSpeed — COMMENTED** [review]
I'd like to see some CEL validations added (the example API group has some examples) and then some integration tests checking these validations please (example API group again). Let me know if you want any pointers on this

**Comment by adriengentil** [comment]
> I'd like to see some CEL validations added (the example API group has some examples) and then some integration tests checking these validations please (example API group again). Let me know if you want any pointers on this

I added CELs to prevent the mutation of `CloudControllerManagerState`, and to prevent the add/remove of `CloudControllerManagerStatus` and `CloudControllerManagerState` . I added few tests as well.

**Review by elmiko — COMMENTED** [review]
looks good to me, would like to get @JoelSpeed 's approval

/lgtm

File: config/v1/stable.infrastructure.testsuite.yaml [hunk 5]
```diff
@@ -56,6 +56,120 @@ tests:
...
+        spec: {}
+        status:
+          platform: External
+          platformStatus:
+            type: External
+            external:
+              cloudControllerManager:
+                state: External
+      expectedStatusError: " status.platformStatus.external.cloudControllerManager.state: Invalid value: \"string\": state is immutable once set"
+    - name: Should not be able to add External cloudControllerManager state once it was set
```
Comment (JoelSpeed) [inline config/v1/stable.infrastructure.testsuite.yaml:83]: Do we have an issue with someone adding it to an existing cluster? I thought we allowed someone to add a CCM, but not remove it? 
By adding a CCM, they start benefiting from cloud provider specific details, such as annotations on new nodes for topology, or service of type load balancer.
But removing a CCM is undefined, how would it clean up services and their load balancers.

I think we went to prevent `External` -> `""` or removed, but, we should allow no CCM to start using one IMO

CC @elmiko do you agree?

File: config/v1/types_infrastructure.go [hunk 6]
```diff
@@ -315,8 +315,45 @@ type PlatformSpec struct {
...
+	CloudControllerManagerExternal CloudControllerManagerState = "External"
+
+	// Cloud Controller Manager is disabled and not expected to be installed.
+	// This value indicates that new nodes should not be tainted
+	// and no extra node initialization is expected from the cloud controller manager.
+	CloudControllerManagerNone CloudControllerManagerState = "None"
+)
+
+// CloudControllerManagerStatus holds the state of Cloud Controller Manager (a.k.a. CCM or CPI) related settings
+// +kubebuilder:validation:XValidation:rule="has(self.state) == has(oldSelf.state)",message="state cannot be added or removed once set"
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:334]: I would have expected can only not be removed once set, which would be more like
```suggestion
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.state) || has(self.state)",message="state cannot be removed once set"
```

File: config/v1/types_infrastructure.go [hunk 7]
```diff
@@ -315,8 +315,45 @@ type PlatformSpec struct {
...
+	// state determines whether or not an external Cloud Controller Manager is expected to
+	// be installed within the cluster.
+	// https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+	//
+	// When set to "External", new nodes will be tainted as uninitialized when created,
+	// preventing them from running workloads until they are initialized by the cloud controller manager.
+	// When omitted or set to "None", new nodes will be not tainted
+	// and no extra initialization from the cloud controller manager is expected.
+	// +kubebuilder:validation:Enum="";External;None
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="state is immutable once set"
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:345]: I think we want to allow a transition to External, but not away from it, lets confirm that and then adjust this

File: config/v1/stable.infrastructure.testsuite.yaml (hunk 5 above)
Comment (adriengentil) [reply config/v1/stable.infrastructure.testsuite.yaml:83]: My understanding was to deny any transitions, transitioning from "None" to "External" would also mean that the existing nodes will never be initialized by the CCM unless we uninitialized them manually first (set the label `node.cloudprovider.kubernetes.io/uninitialized`).

It's a behavior we observed when we made tests with Oracle Cloud's CCM.

Comment (elmiko) [reply config/v1/stable.infrastructure.testsuite.yaml:83]: this is a good question. my concern is that even with a None->External transition there would still need to be a roll out of the core kube components to adjust the `--cloud-provider` flag, and the user would need to coordinate the deployment of their CCM.

@JoelSpeed what is the official API stance on making this immutable now, but changing the mutability later if we get requests from users to enable the transition behavior, could we do that?

my preference would be to keep this immutable for now until we hear that people actually want to transition from None->External.

File: config/v1/types_infrastructure.go (hunk 7 above)
Comment (elmiko) [reply config/v1/types_infrastructure.go:345]: i left some comments above about this

File: config/v1/stable.infrastructure.testsuite.yaml (hunk 5 above)
Comment (mhrivnak) [reply config/v1/stable.infrastructure.testsuite.yaml:83]: Making it immutable now and enabling the transition later should be about the same amount of work as enabling the transition now. I can't think of a downside of deferring that part, other than we have to change the validation and documentation later.

**Comment by JoelSpeed** [comment]
I'm willing to approve this based on the caveats that:
* We are not expecting in the first pass anyone to be able to change the state of their cluster CCM post install
* If we find this becomes an issue, we reserve the right to relax later to allow to go from no CCM to CCM but not the other direction
* No one is actually using external right now since it's new
I would prefer if we can see some QE ack that this is working in payload and not breaking anything existing before we merge though, @adriengentil have you organised a PR that bumps the API in whichever component installs this, that QE could use to test out the API and the validations here?

**Comment by adriengentil** [comment]
> @adriengentil have you organised a PR that bumps the API in whichever component installs this, that QE could use to test out the API and the validations here?

Opened this one against KCMO: https://github.com/openshift/cluster-kube-controller-manager-operator/pull/736
Let me know if that's enough.

BTW, I noticed we will need the featuregate PRs merged for [MCO](https://github.com/openshift/machine-config-operator/pull/3688) and [CCCMO](https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/249) before I can re-vendor the api and library-go in these repos.

**Comment by JoelSpeed** [comment]
> Opened this one against KCMO: https://github.com/openshift/cluster-kube-controller-manager-operator/pull/736
Let me know if that's enough.

I don't think it is, I'm not sure how the infrastructure CRD gets installed into the cluster, but we need to work out how it does, so that we can ensure that the correct repository is updated

**Comment by huali9** [comment]
@adriengentil I build image using the pr https://github.com/openshift/cluster-kube-controller-manager-operator/pull/736, then install a cluster setting CCM External like this
```
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
  creationTimestamp: null
  name: cluster
spec:
  cloudConfig:
    name: ""
  platformSpec:
    type: External
status:
  apiServerInternalURI: https://api-int.huliu-vs17c.qe.devcluster.openshift.com:6443
  apiServerURL: https://api.huliu-vs17c.qe.devcluster.openshift.com:6443
  controlPlaneTopology: HighlyAvailable
  cpuPartitioning: None
  etcdDiscoveryDomain: ""
  infrastructureName: huliu-vs17c-ftdbk
  infrastructureTopology: HighlyAvailable
  platform: External  
  platformStatus:
    type: External
    external:
      cloudControllerManager:
        state: External 
```
, the cluster install successfull, and seems CCM External doesn't take effect. Because I checked the infrastructure on the cluster, it's
```
liuhuali@Lius-MacBook-Pro huali-test % oc get infrastructure cluster -oyaml
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
  creationTimestamp: "2023-05-17T03:43:22Z"
  generation: 1
  name: cluster
  resourceVersion: "514"
  uid: 96ec2307-546d-4e3e-98b6-c6b4806e0d92
spec:
  cloudConfig:
    name: ""
  platformSpec:
    type: External
status:
  apiServerInternalURI: https://api-int.huliu-vs17c.qe.devcluster.openshift.com:6443
  apiServerURL: https://api.huliu-vs17c.qe.devcluster.openshift.com:6443
  controlPlaneTopology: HighlyAvailable
  etcdDiscoveryDomain: ""
  infrastructureName: huliu-vs17c-ftdbk
  infrastructureTopology: HighlyAvailable
  platform: External
  platformStatus:
    external: {}
    type: External
```
Can you please take a look? Thanks!

**Comment by huali9** [comment]
Thanks @adriengentil I build image using the pr https://github.com/openshift/cluster-config-operator/pull/306, then install a cluster setting CCM External, the configuration takes effect now, but the cluster installed successfully, and there is no such label `node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule` on the nodes. Can you please take a look? Thanks!
```
liuhuali@Lius-MacBook-Pro huali-test % oc get infrastructure cluster -oyaml                                   
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
  creationTimestamp: "2023-05-18T01:38:11Z"
  generation: 1
  name: cluster
  resourceVersion: "517"
  uid: 4a0a7ef0-505d-4fd3-82c5-da9bc000a99d
spec:
  cloudConfig:
    name: ""
  platformSpec:
    type: External
status:
  apiServerInternalURI: https://api-int.huliu-vs18c.qe.devcluster.openshift.com:6443
  apiServerURL: https://api.huliu-vs18c.qe.devcluster.openshift.com:6443
  controlPlaneTopology: HighlyAvailable
  etcdDiscoveryDomain: ""
  infrastructureName: huliu-vs18c-dh8zg
  infrastructureTopology: HighlyAvailable
  platform: External
  platformStatus:
    external:
      cloudControllerManager:
        state: External
    type: External
liuhuali@Lius-MacBook-Pro huali-test % oc get node --show-labels           
NAME                                STATUS   ROLES                  AGE   VERSION           LABELS
huliu-vs18c-dh8zg-compute-0         Ready    worker                 82m   v1.27.1+20a4409   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=huliu-vs18c-dh8zg-compute-0,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos
huliu-vs18c-dh8zg-compute-1         Ready    worker                 82m   v1.27.1+20a4409   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=huliu-vs18c-dh8zg-compute-1,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos
huliu-vs18c-dh8zg-control-plane-0   Ready    control-plane,master   92m   v1.27.1+20a4409   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=huliu-vs18c-dh8zg-control-plane-0,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,node-role.kubernetes.io/master=,node.openshift.io/os_id=rhcos
huliu-vs18c-dh8zg-control-plane-1   Ready    control-plane,master   92m   v1.27.1+20a4409   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=huliu-vs18c-dh8zg-control-plane-1,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,node-role.kubernetes.io/master=,node.openshift.io/os_id=rhcos
huliu-vs18c-dh8zg-control-plane-2   Ready    control-plane,master   92m   v1.27.1+20a4409   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=huliu-vs18c-dh8zg-control-plane-2,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,node-role.kubernetes.io/master=,node.openshift.io/os_id=rhcos
liuhuali@Lius-MacBook-Pro huali-test % oc get deploy -n openshift-cloud-controller-manager
No resources found in openshift-cloud-controller-manager namespace.
liuhuali@Lius-MacBook-Pro huali-test % oc get pod -n openshift-cloud-controller-manager
No resources found in openshift-cloud-controller-manager namespace.
```

**Comment by adriengentil** [comment]
Thanks for the test @huali9, the nodes are not labelled because we need my change opend in [library-go](https://github.com/openshift/library-go/pull/1496) to be re-vendored in few other components (MCO, KCMO, and CCCMO) to be re-vendored.
Can you at least test that `status.platformStatus.external.cloudControllerManager.state` is immutable? (e.g: it cannot be updated to `None` or to an empty string)

@JoelSpeed @elmiko do you see other tests that could be performed?

BTW, the CI passed in https://github.com/openshift/cluster-config-operator/pull/306 so I guess this change does not cause any kind of regressions.

**Comment by huali9** [comment]
>Can you at least test that status.platformStatus.external.cloudControllerManager.state is immutable? (e.g: it cannot be updated to None or to an empty string)

Thanks @adriengentil Yes, I tested that, after change the state from `External` to `None`, then check it's still `External`.

**Comment by JoelSpeed** [comment]
/approve
/hold This feature is currently blocked on openshift/machine-config-operator#3688, once that has merged, we can merge the API and line up other PRs to merge in quick succession.
If merge requirements are relaxed for the 4.14 main branch we can merge the API without the blocker and queue up the other PRs

**Comment by adriengentil** [comment]
> This feature is currently blocked on https://github.com/openshift/machine-config-operator/pull/3688, once that has merged, we can merge the API and line up other PRs to merge in quick succession.

This feature is also dependent on https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/252

**Review by adriengentil — COMMENTED** [review]
Added `CloudControllerManagerStatus` behind `TechPreviewNoUpgrade`, I have few questions

File: config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml [hunk 8]
```diff
@@ -593,6 +593,9 @@ spec:
                     external:
                       description: External contains settings specific to the generic External infrastructure provider.
                       type: object
+                      x-kubernetes-validations:
+                        - message: cloudControllerManager added or removed once set
+                          rule: has(self.cloudControllerManager) == has(oldSelf.cloudControllerManager)
```
[the remark this answers is not in the document]
Comment (elmiko) [reply config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml:598]: since we are putting this behind a feature gate for testing, it seems like removing the validation for the moment would be accepable.

File: config/v1/feature_gates.go [hunk 9]
```diff
@@ -83,6 +83,16 @@ var (
 		OwningProduct:       ocpSpecific,
 	}
 
+	FeatureGateExternalCloudProviderExternal = FeatureGateName("ExternalCloudProviderExternal")
+	externalCloudProviderExternal            = FeatureGateDescription{
+		FeatureGateAttributes: FeatureGateAttributes{
+			Name: FeatureGateExternalCloudProviderExternal,
+		},
+		OwningJiraComponent: "cloud-provider",
+		ResponsiblePerson:   "jspeed",
```
[the remark this answers is not in the document]
Comment (elmiko) [reply config/v1/feature_gates.go:92]: technically @JoelSpeed is the team lead, but if he is happy with having me here then i am too =)

**Review by elmiko — COMMENTED** [review]
seems ok to me, just need to get a review from @JoelSpeed

File: config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml:598]: If you can wait until the end of next week, this issue has already been raise and I intend to work during shift week to find a solution for this. My intention is to allow CEL validations to be gated by a featureset, as has been done for fields themselves

File: config/v1/feature_gates.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply config/v1/feature_gates.go:92]: You can put mike there, then if anything blows up it's his responsibility not mine 😉

File: config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply config/v1/0000_10_config-operator_01_infrastructure-Default.crd.yaml:598]: Spoke with Mike about this, I think we can drop this particular validation for now and re-introduce later.
It is safe to introduce later since it is a transitional validation, ie, no matter the data that exists in etcd, it won't cause an invalid data to be stored, also, as this is tech preview, we can adjust before we make it GA

File: config/v1/feature_gates.go [hunk 10]
```diff
@@ -83,6 +83,16 @@ var (
 		OwningProduct:       ocpSpecific,
 	}
 
+	FeatureGateExternalCloudProviderExternal = FeatureGateName("ExternalCloudProviderExternal")
```
Comment (JoelSpeed) [inline config/v1/feature_gates.go:86]: Is this definitely what we want to call this? Seems funny to have `External` twice, not sure it's super important what it's called since we intend to remove it eventually

File: config/v1/types_infrastructure.go [hunk 11]
```diff
@@ -315,8 +315,46 @@ type PlatformSpec struct {
...
+	// +optional
+	State CloudControllerManagerState `json:"state"`
+}
+
 // ExternalPlatformStatus holds the current status of the generic External infrastructure provider.
-type ExternalPlatformStatus struct{}
+// +kubebuilder:validation:XValidation:rule="has(self.cloudControllerManager) == has(oldSelf.cloudControllerManager)",message="cloudControllerManager added or removed once set"
+type ExternalPlatformStatus struct {
+	// CloudControllerManager contains settings specific to the external Cloud Controller Manager (a.k.a. CCM or CPI)
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:353]:
```suggestion
	// cloudControllerManager contains settings specific to the external Cloud Controller Manager (a.k.a. CCM or CPI)
```

File: config/v1/types_infrastructure.go [hunk 12]
```diff
@@ -315,8 +315,46 @@ type PlatformSpec struct {
...
+	// +openshift:enable:FeatureSets=TechPreviewNoUpgrade
+	// +optional
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:355]: Since this is optional, you should document in the godoc what happens when the field is omitted

File: config/v1/types_infrastructure.go [hunk 13]
```diff
@@ -315,8 +315,46 @@ type PlatformSpec struct {
...
+)
+
+// CloudControllerManagerStatus holds the state of Cloud Controller Manager (a.k.a. CCM or CPI) related settings
+// +kubebuilder:validation:XValidation:rule="has(self.state) == has(oldSelf.state)",message="state cannot be added or removed once set"
+type CloudControllerManagerStatus struct {
+	// state determines whether or not an external Cloud Controller Manager is expected to
+	// be installed within the cluster.
+	// https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+	//
+	// When set to "External", new nodes will be tainted as uninitialized when created,
```
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:340]: Should have a line in here dictating the valid values
```suggestion
	// Valid values are "External", "None" and omitted.
	// When set to "External", new nodes will be tainted as uninitialized when created,
```

**Review by JoelSpeed — COMMENTED** [review]
Has this been E2E tested at all? Looking at the API, as `ExternalPlatformStatus` is behind a pointer, this isn't going to break anything at the moment, but if someone added a new field, then the JSON representation marshalling would include the empty fields for `cloudControllerManager` and `state`, which would then fail the current validations.

If instead the CEL on `CloudControllerManagerStatus` checked that if state was new that the value was empty, that would be more future proof
```
// +kubebuilder:validation:XValidation:rule="(has(self.state) == has(oldSelf.state)) || self.state == \"\""
```
This way it would allow a json marshal to marshal the fields without breaching the CEL and without changing the semantics of what is defined

Basically I think 
```
external: {}
```
to 
```
external:
  cloudControllerManager:
    state: ""
```
Should be allowed since syntactically it makes no difference to the cluster

**Comment by elmiko** [comment]
> Has this been E2E tested at all?

i'm not sure we have an e2e for it yet, but @rvanderp3 and @mtulio have been investigating, perhaps they have a comment here.

**Comment by JoelSpeed** [comment]
> If instead the CEL on CloudControllerManagerStatus checked that if state was new that the value was empty, that would be more future proof
> ```
> // +kubebuilder:validation:XValidation:rule="(has(self.state) == has(oldSelf.state)) || self.state == \"\""
> ```

Coming back to this thought, all we actually care about is that it doesn't go from empty to External, so the transition rule really should either allow it to stay as is, or allow it to transition not to External

```
// +kubebuilder:validation:XValidation:rule="(has(self.state) == has(oldSelf.state)) || (!has(oldSelf.state) && self.state != \"External\"")
```

This makes the following IIUC (where `-` means it's not there):
| old | new | allowed |
| --- | --- | --- |
| - | - | true |
| External | External | true |
| None | None | true |
| "" | "" | true |
| - | "" | true |
| - | None | true |
| - | External | false |
| "" | - | false |
| None | - | false |
| External | - | false |

If there's ever a change of transition between values while both are present the immutable validation on the field itself will catch that

**Comment by mtulio** [comment]
> > Has this been E2E tested at all?
> 
> i'm not sure we have an e2e for it yet, but @rvanderp3 and @mtulio have been investigating, perhaps they have a comment here.

I'm not quite sure we have. recovering from the EP I found [this Q/A](https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/infrastructure-external-platform-type.md?plain=1#L485-L488). There is a recommendation to prioritize in-repo validations, I also quickly looked at the `origin` repo and didn't find anything related, so I would assume the e2e workflows we have currently running in for Platform External in vSphere platform is covering the same e2e of Platform None.

Furthermore, inspecting the results published by OPCT on executions in OCP 4.13.2 for Platform None[1A] VS External[2], (values below), I can see similar counters for skips and failures

[1] - Platform None (A)
~~~
> OpenShift Provider Certification Summary <

 Kubernetes API Server version		: v1.26.5+0001a21
 OpenShift Container Platform version	: 4.13.2
 - Cluster Update Progressing		: False
 - Cluster Target Version		: Cluster version is 4.13.2
						
 OCP Infrastructure:			
 - PlatformType				: None
 - Name					: ci-op-7vi5vvjm-56bd3-zqtp6
						
 Plugins summary by name:		  Status [Total/Passed/Failed/Skipped] (timeout)
 - 10-openshift-kube-conformance	: failed [724/695/29/0] (0)
 - 20-openshift-conformance-validated	: failed [3808/1681/53/2074] (0)
									
 Health summary:			  [A=True/P=True/D=True]	
 - Cluster Operators			: [33/0/0]
 - Node health				: 6/6  (100%)
 - Pods health				: 249/249  (100%)

> Processed Summary <

 Total tests by conformance suites:
 - kubernetes/conformance: 364 
 - openshift/conformance: 3443 

 Result Summary by conformance plugins:
 - 10-openshift-kube-conformance:
   - Status: failed
   - Total: 724
   - Passed: 695
   - Failed: 29
   - Timeout: 0
   - Skipped: 0
   - Failed (without filters) : 29
   - Failed (Filter SuiteOnly): 0
   - Failed (Filter CI Flakes): 0
   - Status After Filters     : pass
 - 20-openshift-conformance-validated:
   - Status: failed
   - Total: 3808
   - Passed: 1681
   - Failed: 53
   - Timeout: 0
   - Skipped: 2074
   - Failed (without filters) : 53
   - Failed (Filter SuiteOnly): 21
   - Failed (Filter CI Flakes): 4
   - Status After Filters     : failed

~~~

- Platform External

~~~
> OpenShift Provider Certification Summary <

 Kubernetes API Server version		: v1.26.5+0001a21
 OpenShift Container Platform version	: 4.13.2
						
 OCP Infrastructure:			
 - PlatformType				: External (vSphere)
 - Name					: ci-op-3s74i69p-ff62f-tq66h
				
 Plugins summary by name:		  Status [Total/Passed/Failed/Skipped] (timeout)
 - 10-openshift-kube-conformance	: failed [723/694/29/0] (0)
 - 20-openshift-conformance-validated	: failed [3811/1677/60/2074] (0)
									
 Health summary:			  [A=True/P=True/D=True]	
 - Cluster Operators			: [33/0/0]
 - Node health				: 6/6  (100%)
 - Pods health				: 246/246  (100%)

> Processed Summary <

 Result Summary by conformance plugins:
 - 10-openshift-kube-conformance:
   - Status: failed
   - Total: 723
   - Passed: 694
   - Failed: 29
   - Timeout: 0
   - Skipped: 0
   - Failed (without filters) : 29
   - Failed (Filter SuiteOnly): 0
   - Failed (Filter CI Flakes): 0
   - Status After Filters     : pass
 - 20-openshift-conformance-validated:
   - Status: failed
   - Total: 3811
   - Passed: 1677
   - Failed: 60
   - Timeout: 0
   - Skipped: 2074
   - Failed (without filters) : 60
   - Failed (Filter SuiteOnly): 24
   - Failed (Filter CI Flakes): 3
   - Status After Filters     : failed

~~~

I checked another None execution and it's similar[1]

~~~
> OpenShift Provider Certification Summary <

 Kubernetes API Server version		: v1.26.5+0001a21
 OpenShift Container Platform version	: 4.13.2
						
 OCP Infrastructure:			
 - PlatformType				: None
 - Name					: ci-op-7s946tcy-1b8d1-2j6dp

 Plugins summary by name:		  Status [Total/Passed/Failed/Skipped] (timeout)
 - 10-openshift-kube-conformance	: failed [726/695/31/0] (0)
 - 20-openshift-conformance-validated: failed [3811/1679/58/2074] (0)

(...)
 - 20-openshift-conformance-validated:
   - Status: failed
   - Total: 3811
   - Passed: 1679
   - Failed: 58
   - Timeout: 0
   - Skipped: 2074
   - Failed (without filters) : 58
   - Failed (Filter SuiteOnly): 23
   - Failed (Filter CI Flakes): 3
   - Status After Filters     : failed

~~~

File: config/v1/feature_gates.go (hunk 10 above)
Comment (adriengentil) [reply config/v1/feature_gates.go:86]: I agree it's not ideal, I kept the same naming scheme as `FeatureGateExternalCloudProviderGCP` and `FeatureGateExternalCloudProviderAzure` to stay consistent. Otherwise, something like `FeatureGateCloudControllerState` ?

Comment (elmiko) [reply config/v1/feature_gates.go:86]: agree that the stutter isn't great, but since it follows the others i think it can be acceptable

**Review by elmiko — COMMENTED** [review]
/lgtm

/hold

@adriengentil feel free to remove the hold once https://github.com/openshift/cluster-config-operator/pull/306 is passing the e2e-aws-ovn-techpreview test, or we can determine it is not failing due to this change.

**Comment by adriengentil** [comment]
/unhold

all tests passed in https://github.com/openshift/cluster-config-operator/pull/306
